jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java

```diff
@@ -114,64 +117,89 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
+            writer.append("<!DOCTYPE html>\n");
+            writer.append("<html lang=\"en\">\n<head>\n");
+            writer.append("<title>Error 404 - Not Found</title>\n");
+            writer.append("<meta charset=\"utf-8\">\n");
+            writer.append("<style>body { font-family: sans-serif; } table, td { border: 1px solid #333; } td, th { padding: 5px; } thead, tfoot { background-color: #333; color: #fff; } </style>\n");
+            writer.append("</head>\n<body>\n");
+            writer.append("<h2>Error 404 - Not Found.</h2>\n");
+            writer.append("<p>No context on this server matched or handled this request.</p>\n");
+            writer.append("<p>Contexts known to this server are:</p>\n");
 
             Server server = getServer();
             Handler[] handlers = server==null?null:server.getChildHandlersByClass(ContextHandler.class);
 
+            writer.append("<table class=\"contexts\"><thead><tr>");
+            writer.append("<th>Context Path</th>");
+            writer.append("<th>Display Name</th>");
+            writer.append("<th>State</th>");
+            writer.append("<th>LifeCycle</th>");
+            writer.append("</tr></thead><tbody>\n");
+
             for (int i=0;handlers!=null && i<handlers.length;i++)
             {
+                writer.append("<tr><td>");
+                // Context Path
                 ContextHandler context = (ContextHandler)handlers[i];
+
+                StringBuilder href = new StringBuilder();
+
+                String contextPath = context.getContextPath();
+                href.append(contextPath);
+                if (contextPath.length() > 1 && !contextPath.endsWith("/"))
+                {
+                    href.append("/");
+                }
+
                 if (context.isRunning())
                 {
-                    writer.write("<li><a href=\"");
-                    if (context.getVirtualHosts()!=null && context.getVirtualHosts().length>0)
-                        writer.write(request.getScheme()+"://"+context.getVirtualHosts()[0]+":"+request.getLocalPort());
-                    writer.write(context.getContextPath());
-                    if (context.getContextPath().length()>1 && context.getContextPath().endsWith("/"))
-                        writer.write("/");
-                    writer.write("\">");
-                    writer.write(context.getContextPath());
-                    if (context.getVirtualHosts()!=null && context.getVirtualHosts().length>0)
-                        writer.write("&nbsp;@&nbsp;"+context.getVirtualHosts()[0]+":"+request.getLocalPort());
-                    writer.write("&nbsp;--->&nbsp;");
-                    writer.write(context.toString());
-                    writer.write("</a></li>\n");
+                    writer.append("<a href=\"").append(href).append("\">");
                 }
-                else
+                writer.append(context.getContextPath());
```

Comment:
Use the `contextPath` variable defined above.

Reply:
But when writing the Textual form (not in a URL) then we should probably escape only the `%` HTML entity to `&#37;`, all other characters must be safe for that contextPath to even work properly.
In fact, we should probably do a proper URL safety check on `ContextHandler.setContextPath(String)` to ensure it doesn't use unsafe characters.